File: src/main/java/work/lclpnet/ap2/impl/game/item/SpecialItemRegistry.java
package work.lclpnet.ap2.impl.game.item;

import org.json.JSONObject;
import work.lclpnet.ap2.impl.ds.WeightedList;

import java.util.HashMap;
import java.util.Map;

public class SpecialItemRegistry implements SpecialItemRegistrar {

    private final Map<String, Entry> items = new HashMap<>();

    @Override
    public SpecialItemRegistrar register(SpecialItem item, float chance) {
        if (chance < 0) throw new IllegalArgumentException("Chance cannot be negative");

        items.put(item.id(), new Entry(item, chance));

        return this;
    }

    public WeightedList<SpecialItem> weightedItems(JSONObject overrides) {
        var weighted = new WeightedList<SpecialItem>(items.size());

        for (var registered : items.entrySet()) {
            String key = registered.getKey();
            Entry entry = registered.getValue();
            float chance = overrides.optFloat(key);

            if (Float.isNaN(chance)) {
                chance = entry.chance;
            }

            if (chance > 0) {
                weighted.add(entry.item, entry.chance);
            }
        }

        return weighted;
    }

    private record Entry(SpecialItem item, float chance) {}
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/item/SpecialItemRegistry.kt
package work.lclpnet.ap2.game.item

import com.google.common.collect.Iterables
import org.json.JSONObject
import work.lclpnet.gaco.ds.WeightedList

class SpecialItemRegistry : SpecialItemRegistrar {

    private val items = HashMap<String, Entry>()

    override fun register(item: SpecialItem, chance: Float): SpecialItemRegistrar {
        require(chance >= 0) { "Chance cannot be negative" }

        items[item.id] = Entry(item, chance)

        return this
    }

    fun weightedItems(overrides: JSONObject): WeightedList<SpecialItem> {
        val weighted = WeightedList<SpecialItem>(items.size)

        for ((key, entry) in items) {
            var chance = overrides.optFloat(key)

            if (chance.isNaN()) {
                chance = entry.chance
            }

            if (chance > 0) {
                weighted.add(entry.item, entry.chance)
            }
        }

        return weighted
    }

    operator fun get(id: String): SpecialItem? =
        items.getOrDefault(id, null)?.item

    fun entries(): Iterable<SpecialItem> =
        Iterables.transform(items.values, Entry::item)

    private data class Entry(
        val item: SpecialItem,
        val chance: Float,
    )
}
